text/0000-name-resolution.md

```diff
@@ -0,0 +1,649 @@
+able to decide if names are accessible. This requires doing privacy checking as
+required during parsing/expansion/import resolution. We can keep the current
+algorithm, but check accessibility on demand, rather than as a separate pass.
+
+During macro expansion, once a name is resolvable, then we can safely perform
+privacy checking, because parsing and macro expansion will never remove items,
+nor change the module structure of an item once it has been expanded.
+
+### Metadata
+
+When a crate is packed into metadata, we must also include the binding table. We
+must include private entries due to macros that the crate might export. We don't
+need data for function bodies. For functions which are serialised for
+inlining/monomorphisation, we should include local data (although it's probably
+better to serialise the HIR or MIR, then the local bindings are unnecessary).
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It's a lot of work and name resolution is complex, therefore there is scope for
+introducing bugs.
+
+The macro changes are not backwards compatible, which means having a macro
+system 2.0. If users are reluctant to use that, we will have two macro systems
+forever.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Naming rules
+
+We could take a subset of the shadowing changes (or none at all), whilst still
+changing the implementation of name resolution. In particular, we might want to
+discard the explicit/glob shadowing rule change, or only allow items, not
+imported names to shadow.
+
+We could also consider different shadowing rules around namespacing. In the
+'globs and explicit names' rule change, we could consider an explicit name to
+shadow both name spaces and emit a custom error. The example becomes:
+
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub trait Qux;
+}
+
+mod boz {
+    use foo::*;
+    use bar::Qux; // Shadows both name spaces.
+
+    fn f(x: &Qux) {   // bound to bar::Qux.
+        let _ = Qux;  // ERROR, unresolved name Qux; the compiler would emit a
+                      // note about shadowing and namespaces.
+    }
+}
+```
+
+## Import resolution algorithm
+
+Rather than lookup names for imports during the fixpoint iteration, one could
+save links between imports and definitions. When lookup is required (for macros,
+or later in the compiler), these links are followed to find a name, rather than
+having the name being immediately available.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Name lookup
+
+The name resolution phase would be replaced by a cut-down name lookup phase,
+where the binding tables generated during expansion are used to lookup names in
+the AST.
+
+We could go further, two appealing possibilities are merging name lookup with
+the lowering from AST to HIR, so the HIR is a name-resolved data structure. Or,
+name lookup could be done lazily (probably with some caching) so no tables
```

Comment:
What's the purpose of doing the lookup lazily? All the names are going to be resolved anyway (unless there's dead code or there are compilation errors, but I don't see much reason to optimize for these cases) then why defer it.